src/ucs/sys/module.c

```diff
@@ -0,0 +1,164 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2001-2018.  ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+
+#define _GNU_SOURCE /* for dladdr(3) */
+
+#include "module.h"
+
+#include <ucs/debug/memtrack.h>
+#include <ucs/debug/assert.h>
+#include <ucs/debug/log.h>
+#include <string.h>
+#include <limits.h>
+#include <dlfcn.h>
+#include <libgen.h>
+
+
+#define UCS_MODULE_PATH_MEMTRACK_NAME   "module_path"
+#define UCS_MODULE_SRCH_PATH_MAX        2
+
+static struct {
+    ucs_init_once_t  init;
+    char             module_ext[NAME_MAX];
+    unsigned         srchpath_cnt;
+    char             *srch_path[UCS_MODULE_SRCH_PATH_MAX];
+} ucs_module_loader_state = {
+    .init         = UCS_INIT_ONCE_INIITIALIZER,
+    .module_ext   = ".so", /* default extension */
+    .srchpath_cnt = 0,
+    .srch_path    = { NULL, NULL}
+};
+
+/* Should be called with lock held */
+static void ucs_module_loader_add_dl_dir()
+{
+    char *dlpath_dup = NULL;
+    size_t max_length;
+    Dl_info dl_info;
+    char *p, *path;
+    int ret;
+
+    (void)dlerror();
+    ret = dladdr((void*)&ucs_module_loader_state, &dl_info);
+    if (ret == 0) {
+        ucs_error("dladdr failed: %s", dlerror());
+        return;
+    }
+
+    ucs_debug("ucs library path: %s", dl_info.dli_fname);
+
+    /* copy extension */
+    dlpath_dup = ucs_strdup(dl_info.dli_fname,
+                            UCS_MODULE_PATH_MEMTRACK_NAME);
+    if (dlpath_dup == NULL) {
+        goto out;
```

Comment:
no need to `goto out`, because `dlpath_dup` was not allocated, can just return here